<launch>
    <!-- Global parameters -->
    <param name="/use_sim_time" value="false"/>

    <!-- Camera publisher with message filters -->
    <node name="camera_publisher" pkg="camera_pkg" type="camera_publisher.py" output="screen" ns="camera">
        <param name="camera_info_url" value="file:/home/arm/self_balancing_robot/src/camera_pkg/config/camera_info.yaml"/>
        <param name="device_id" value="0"/>
        <param name="image_width" value="640"/>
        <param name="image_height" value="480"/>
        <param name="fps" value="30"/>
        <!-- Add timestamp tolerance -->
        <param name="timestamp_tolerance" value="0.03"/>
    </node>

    <!-- Image proc with synchronization -->
    <node pkg="image_proc" type="image_proc" name="image_proc" ns="camera">
        <param name="queue_size" value="10"/>
        <param name="approximate_sync" value="true"/>
        <param name="async_mode" value="false"/>
        <param name="max_delay" value="0.1"/>
        <remap from="image_raw" to="image_raw"/>
        <remap from="camera_info" to="camera_info"/>
    </node>

    <!-- IPM node with synchronization -->
    <node pkg="camera_pkg" type="ipm.py" name="ipm_node" ns="camera">
        <param name="approximate_sync" value="true"/>
        <param name="queue_size" value="10"/>
        <param name="slop" value="0.1"/>
        <remap from="image_rect" to="image_rect"/>
        <remap from="camera_info" to="camera_info"/>
        <rosparam command="load" file="/home/arm/self_balancing_robot/src/camera_pkg/config/ipm_node.yaml"/>
        <param name="camera_frame" value="camera_optical_frame"/>
        <param name="ground_frame" value="ground_plane"/>
        <param name="height" value="1.0"/>
        <param name="output_resolution" type="yaml" value="[640, 480]"/>
    </node>

    <!-- Add clock publisher for better synchronization -->
    <node pkg="topic_tools" type="relay" name="clock_relay" args="/clock /clock_sync"/>
</launch>